### DISCUSSION TITLE
Operator and function overloading

### DISCUSSION THREAD

**Discussion by JM1978** [autodiff/autodiff#164] [Q&A] [answered]
Hi Allan,

First, thank you for making this library available!

I've been running into compiler errors which seem to be related to functions and operators not being overloaded to work with autodiff's dual type.  I've been able to "fix" the missing operator errors by forcing the problematic operand to be the dual type with a separate declaration.

However, I am running into a problem now with the std::pow() function, in which one of the arguments is the dual type.

Is there any trick to getting around this problem, apart from overloading the function myself?  Apologies if this is a basic programming question, as I still have a lot to learn in C++.

**Comment by allanleal** [comment, marked as answer]
Hi @JM1978 ,

Could you please post a very minimal example that reproduces your issue? I'll have a look for you. Thanks for your feedback.

Best regards,
Allan

**Reply by JM1978** [reply]
Hi Allan,

Below is a short piece of code that produces as the first error using g++ 7.5.0 on Ubuntu.

main.cpp: In function ‘std::complex<autodiff::forward::Dual<double, double> > Z(double, const Params&)’:
main.cpp:21:74: error: no matching function for call to ‘pow(std::complex<double>, const dual&)’
  std::complex<autodiff::dual> Zcpe = 1.0 / (p.Q * pow((1.0i * omega), p.a)); 


```
#include <complex>
#include <vector>

#include <eigen3/Eigen/Core>
#include <autodiff/forward.hpp>
#include <autodiff/forward/eigen.hpp>


using namespace std::literals::complex_literals;

/* Structure containing the parameters of the CPE impedance function */
struct Params{
	autodiff::dual Q;
	autodiff::dual a;
};

/* Impedance of a constant phase element (CPE) in complex form */
std::complex<autodiff::dual> Z(const double f, const Params &p){
	const double pi = 3.1415926;
	const double omega = 2.0 * pi * f;
	std::complex<autodiff::dual> Zcpe = 1.0 / (p.Q * pow((1.0i * omega), p.a));	
	return Zcpe;
}

/* wrapper function that returns the real component */
autodiff::dual fRe(const double f, const Params &p){
	return Z(f, p).real();
}


int main(void){
	
	const double f = 0.01;	// ordinary frequency (Hz)
        Params p;
        p.Q = 1e-5;
        p.a = 0.95;
	double derivative = autodiff::derivative(fRe, wrt(p.Q), autodiff::forward::at(f, p));
	std::cout << derivative << std::endl;
	
	return 0;	
}
```

Makefile:
```
CXX = g++
#IDIR = /usr/include/eigen3/ /usr/include/autodiff
INC=$(/usr/include/eigen3/) $(/usr/include/autodiff)
INC_PARAMS = $(foreach d, $(INC), -I$d)
CXXFLAGS = -Wall -std=c++17 $(INC_PARAMS)


test: main.o 
	$(CXX) $(CXXFLAGS) -o test main.o
		
# pattern rule
%.o: %.c
	$(CXX) $(CXXFLAGS) -c $< -o $@

clean:
	rm test main.o 
```

**Reply by allanleal** [reply]
Please format your code when posting. Have a look at this example using `std::complex` and see if it helps:
https://autodiff.github.io/tutorials/#derivatives-of-a-single-variable-function-using-a-custom-scalar-complex

**Reply by JM1978** [reply]
Hi Allan,

That example you linked to helps a lot.

Regarding formatting of code, do you mean I should enable syntax highlighting, or something else?

Thank you

Jason

**Reply by allanleal** [reply]
Hi Jason,

That's good to hear. As to formatting, yes, using markdown (see this [guide](https://guides.github.com/features/mastering-markdown/)).

Please mark this discussion as resolved (if it was resolved indeed!).

Thanks,
Allan
